backend/main.go

```diff
@@ -79,22 +83,41 @@ func main() {
 		IdleTimeout:  60 * time.Second,
 	}
 
-	go func() {
-		err := srv.ListenAndServe()
-		if err == http.ErrServerClosed {
-			err = nil
+	terminationCtx := graceful.WithTermination(context.Background())
+
+	defer graceful.Shutdown(terminationCtx, func(err error, exitCode int) {
+		shutdownCtx := context.WithoutCancel(terminationCtx)
+
+		if wait > 0 {
+			timeoutCtx, cancel := context.WithTimeout(shutdownCtx, wait)
+			defer cancel()
+			shutdownCtx = timeoutCtx
+		}
+
+		if shutdownErr := srv.Shutdown(shutdownCtx); shutdownErr != nil {
+			err = errors.Join(err, fmt.Errorf("server shutdown failed: %w", shutdownErr))
 		}
+
 		if err != nil {
-			log.Errorln(err)
+			log.Errorln("Shutdown errors:", err)
+			os.Exit(exitCode)
+		} else {
+			log.Infoln("Server gracefully stopped")
+		}
+	})
+
+	go func() {
+		if err := srv.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Errorln("Server error:", err)
+			graceful.Terminate(terminationCtx, err, 1)
 		}
 	}()
 
 	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt)
-	<-c
-	ctx, cancel := context.WithTimeout(context.Background(), wait)
-	defer cancel()
-	srv.Shutdown(ctx)
-	log.Infoln("Shutting down")
-	os.Exit(0)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```

Comment:
You don't need to listen the system signals directly because we already listen them [under the hood of graceful pkg](https://github.com/werf/common-go/blob/main/pkg/graceful/graceful.go#L46). 

Check one mote time my previous comment with code snippet.

Please, keep it simple =)